pkg/controller/installation/validation.go

```diff
@@ -145,6 +148,37 @@ func validateCustomResource(instance *operatorv1.Installation) error {
 		}
 	}
 
+	switch instance.Spec.CNI.Type {
+	case operatorv1.PluginCalico:
+	case operatorv1.PluginGKE:
+		switch instance.Spec.KubernetesProvider {
+		case operatorv1.ProviderEKS:
+			fallthrough
+		case operatorv1.ProviderAKS:
+			return fmt.Errorf("kubernetesProvider of %s is invalid with CNI.type %s",
+				instance.Spec.KubernetesProvider, instance.Spec.CNI.Type)
+		}
+	case operatorv1.PluginAmazonVPC:
+		switch instance.Spec.KubernetesProvider {
+		case operatorv1.ProviderGKE:
+			fallthrough
+		case operatorv1.ProviderAKS:
+			return fmt.Errorf("kubernetesProvider of %s is invalid with CNI.type %s",
+				instance.Spec.KubernetesProvider, instance.Spec.CNI.Type)
+		}
+	case operatorv1.PluginAzureVNET:
+		switch instance.Spec.KubernetesProvider {
+		case operatorv1.ProviderGKE:
+			fallthrough
+		case operatorv1.ProviderEKS:
+			return fmt.Errorf("kubernetesProvider of %s is invalid with CNI.type %s",
+				instance.Spec.KubernetesProvider, instance.Spec.CNI.Type)
+		}
+	default:
+		return fmt.Errorf("Invalid CNI.type %s, it should be one of %s",
+			instance.Spec.CNI.Type, strings.Join(operatorv1.CNIPluginTypesString, ","))
```

Comment:
if this is the only place `CNIPluginTypesString` is used, why not just put the actual strings here?

Reply:
Because then the strings to will be defined 2 places, once where the type is defined and then also here.
With the current way it will be more obvious that a change should be made when adding a new PluginType. If we just listed them here then when adding a new PluginType it wouldn't be obvious that this code should be updated too.